### PR TITLE
Add experimental support for non-x86 archs and MacOS

### PR DESCRIPTION
**Pull request by IonAgorria**
This PR makes x86 specifics optional in case the platform is different to allow compiling it under ARM64 (not tested in other archs)

Also adds changes required to make it compile under MacOS and run a D3D9 using game.

Please review my changes carefully as this is not my area of expertise and there may be better ways to do this.

### CODE REVIEW

**Comment by Joshua-Ashton** [comment]
Looks fine to me, no comment on the Mac-specific stuff though.

**Comment by IonAgorria** [comment]
I can split the Mac-specific stuff into another PR if its better.

**Comment by Joshua-Ashton** [comment]
Should be fine.

File: meson.build [hunk 1]
```diff
@@ -58,6 +58,13 @@ else
   dxvk_library_path = meson.source_root() + '/lib32'
 endif
 
+# Set define for arch
+if (cpu_family == 'x86' or cpu_family == 'x86_64')
+  add_project_arguments('-DCPU_ARCH_X86', language : ['c', 'cpp'])
```
Comment (Joshua-Ashton) [inline meson.build:63]: Probably define both x86_64 too?

Or rather, why not just use ``__x86_64__`` and friends?

**Review by misyltoad — CHANGES_REQUESTED** [review]

File: src/util/sync/sync_spinlock.h [hunk 2]
```diff
@@ -21,7 +21,11 @@ namespace dxvk::sync {
   void spin(uint32_t spinCount, const Fn& fn) {
     while (unlikely(!fn())) {
       for (uint32_t i = 1; i < spinCount; i++) {
+#if defined(CPU_ARCH_X86)
         _mm_pause();
+#elif defined(CPU_ARCH_ARM)
+        __asm__ __volatile__ ("yield");
```
Comment (Joshua-Ashton) [inline src/util/sync/sync_spinlock.h:27]: Is there an intrinsic for this?

File: src/util/util_vector.h [hunk 3]
```diff
@@ -151,10 +151,17 @@ namespace dxvk {
 
   inline Vector4 replaceNaN(Vector4 a) {
     Vector4 result;
+#ifdef CPU_ARCH_X86
     __m128 value = _mm_load_ps(a.data);
     __m128 mask  = _mm_cmpeq_ps(value, value);
            value = _mm_and_ps(value, mask);
     _mm_store_ps(result.data, value);
+#else
+    if (a[0] != a[0]) a[0] = 0;
```
Comment (Joshua-Ashton) [inline src/util/util_vector.h:160]: 0.0f

Why not do a loop here too?

**Review by misyltoad — CHANGES_REQUESTED** [review]

File: include/native/wsi/native_sdl2.h [hunk 4]
```diff
@@ -1,7 +1,13 @@
 #include <cstdint>
 #include <windows.h>
 
+#ifdef DXVK_PLATFORM_DARWIN
+#include <SDL.h>
+#include <SDL_vulkan.h>
```
Comment (Joshua-Ashton) [inline include/native/wsi/native_sdl2.h:6]: Why is this needed? Can we solve it in the build script? :(

**Review by misyltoad — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_device.cpp [hunk 5]
```diff
@@ -3865,7 +3869,11 @@ namespace dxvk {
     enabled.core.features.sampleRateShading = VK_TRUE;
     enabled.core.features.samplerAnisotropy = supported.core.features.samplerAnisotropy;
     enabled.core.features.shaderClipDistance = VK_TRUE;
+#ifdef DXVK_PLATFORM_DARWIN
+    enabled.core.features.shaderCullDistance = supported.core.features.shaderCullDistance;
+#else
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:3874]: I dont think we even use this, so feel free to just do this anyway.

File: src/d3d9/d3d9_device.cpp [hunk 6]
```diff
@@ -3836,7 +3836,11 @@ namespace dxvk {
     DxvkDeviceFeatures enabled = {};
 
     // Geometry shaders are used for some meta ops
+#ifdef DXVK_PLATFORM_DARWIN
+    enabled.core.features.geometryShader = supported.core.features.geometryShader;
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_device.cpp:3840]: We only use this for swvp ProcessVertices so I dont mind just doing this always.

**Review by misyltoad — CHANGES_REQUESTED** [review]

File: meson.build (hunk 1 above)
Comment (IonAgorria) [reply meson.build:63]: I rather prefer meson to provide us the arch to target than having to look each compiler define's for archs.
Also the defines are used only to have x86/x86_64 or ARM/Aarch64 specific code, not between 32 or 64 bits.

File: src/util/sync/sync_spinlock.h (hunk 2 above)
Comment (IonAgorria) [reply src/util/sync/sync_spinlock.h:27]: Looked up GCC intrinsic but there is none for ARM/64 :( apparently only MSVC has "__yield()"...

Also after the loop it calls dxvk::this_thread::yield(); which calls stdlib yield, isn't already doing same as the intrinsic/asm instructions under the hood? I wasn't sure anyway so I just searched the x86 pause equivalent for ARMs and came with yield

File: src/util/util_vector.h (hunk 3 above)
Comment (IonAgorria) [reply src/util/util_vector.h]: Fixed, I thought unrolled loop would be better

File: include/native/wsi/native_sdl2.h (hunk 4 above)
Comment (IonAgorria) [reply include/native/wsi/native_sdl2.h:5]: Interestingly the recommended way is using `#include "SDL.h"` according to FAQ https://wiki.libsdl.org/FAQDevelopment so this way is more correct.

Using SDL2/... didn't work on MacOS, without SDL2/... works on my Linux and MacOS after using dependency() meson call as in MacOS it can't find SDL2 when using find_library(), it also seems the recommended way to search dependencies.

I merged darwin and linux branches in meson for dependencies.

File: src/d3d9/d3d9_device.cpp (hunk 5 above)
Comment (IonAgorria) [reply src/d3d9/d3d9_device.cpp]: Done

File: src/d3d9/d3d9_device.cpp (hunk 6 above)
Comment (IonAgorria) [reply src/d3d9/d3d9_device.cpp:3839]: Done, as a sidenote it seems that bump/normal mapping doesn't work at MacOS in M1 when testing the game but Im not sure if its because of MoltenVK side or because we don't have these 2 features available... anyway is not important as I disabled in game options when targeting this platform.

**Comment by caiiiycuk** [comment]
Useful PR I also need this :+1:

**Comment by q4a** [comment]
Nice) Arm/Aarch64 support is what I need.

File: include/native/wsi/native_sdl2.h (hunk 4 above)
Comment (eli-schwartz) [reply include/native/wsi/native_sdl2.h:5]: Right, the other advantage is that `find_library()` won't let you detect a custom compiled version of SDL2 added to your development environment via PKG_CONFIG_PATH.

Using dependency() also means you can use https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section

**Comment by q4a** [comment]
> We should come back to this after DXVK Native gets upstreamed.

@Joshua-Ashton Upstreamed here? https://github.com/doitsujin/dxvk

**Comment by Joshua-Ashton** [comment]
Yes, I am not super sure how likely it is for Mac stuff to be upstreamed given we cannot really maintain it, but I think the ARM stuff may have a chance. Will see anyway when the time comes.

**Comment by IonAgorria** [comment]
I could remove macos stuff from this PR if desired and maybe make another PR

**Comment by q4a** [comment]
@Joshua-Ashton Is there any chance, that removing macos stuff from this PR will speed up merging?

**Comment by misyltoad** [comment]
Superceded by https://github.com/doitsujin/dxvk/pull/2911
